sources/api/netdog/src/cli/write_resolv_conf.rs

```diff
@@ -1,28 +1,187 @@
 /// Writes /etc/resolv.conf, using DNS API settings if they exist
 pub(crate) struct WriteResolvConfArgs {}
 
+/// A struct representing an interface drop-in that overrides DNS-via-DHCP settings
+#[cfg(net_backend = "systemd-networkd")]
+#[derive(Debug, SystemdUnit)]
+struct InterfaceDNSDropIn {
+    network: Option<NetworkSection>,
+    dhcp4: Option<Dhcp4Section>,
+    dhcp6: Option<Dhcp6Section>,
+    ipv6_accept_ra: Option<IPv6AcceptRASection>,
+}
+
+#[cfg(net_backend = "systemd-networkd")]
+#[derive(Debug, SystemdUnitSection)]
+#[systemd(section = "Network")]
+struct NetworkSection {
+    #[systemd(entry = "DNSDefaultRoute")]
+    dns_default_route: Option<bool>,
+}
+
+#[cfg(net_backend = "systemd-networkd")]
+#[derive(Debug, SystemdUnitSection)]
+#[systemd(section = "DHCPv4")]
+struct Dhcp4Section {
+    #[systemd(entry = "UseDNS")]
+    use_dns: Option<bool>,
+    #[systemd(entry = "UseDomains")]
+    use_domains: Option<bool>,
+}
+
+#[cfg(net_backend = "systemd-networkd")]
+#[derive(Debug, SystemdUnitSection)]
+#[systemd(section = "DHCPv6")]
+struct Dhcp6Section {
+    #[systemd(entry = "UseDNS")]
+    use_dns: Option<bool>,
+    #[systemd(entry = "UseDomains")]
+    use_domains: Option<bool>,
+}
+
+#[cfg(net_backend = "systemd-networkd")]
+#[derive(Debug, SystemdUnitSection)]
+#[systemd(section = "IPv6AcceptRA")]
+struct IPv6AcceptRASection {
+    #[systemd(entry = "UseDNS")]
+    use_dns: Option<bool>,
+    #[systemd(entry = "UseDomains")]
+    use_domains: Option<bool>,
+}
+
 #[cfg(net_backend = "systemd-networkd")]
-fn get_dns_settings(primary_interface: String) -> Result<DnsSettings> {
-    let primary_link_status = NetworkDInterfaceStatus::new(primary_interface)
-        .context(error::NetworkDInterfaceStatusSnafu)?;
-    let dns_settings = DnsSettings::from_config_or_status(&primary_link_status)
-        .context(error::GetDnsSettingsSnafu)?;
-    Ok(dns_settings)
+impl InterfaceDNSDropIn {
+    /// Given API DNS settings create an appropriate drop-in for a network interface.
+    fn new(settings: &DnsSettings, is_primary: bool) -> Self {
+        // Default to not using DNS values from DHCP, and do not use the interface's DNS route to
+        // resolve domains not matching other config.  If the interface is the primary interface,
+        // use DNS API settings to direct the appropriate interface configuration.  This matches
+        // existing wicked behavior by ensuring DNS configuration is sourced from settings and the
+        // primary interface only.
+        let mut should_use_dns_from_dhcp = Some(false);
+        let mut should_use_domains_from_dhcp = Some(false);
+        let mut should_be_dns_default_route = Some(false);
+
+        if is_primary {
+            should_use_dns_from_dhcp = Some(!settings.has_name_servers());
+            should_use_domains_from_dhcp = Some(!settings.has_search_domains());
+            should_be_dns_default_route = Some(!settings.has_name_servers());
```

Comment:
nit: separate bools are fine but maybe do this to indicate that they are intended to be the same value in this path:
```suggestion
            should_be_dns_default_route = should_use_dns_from_dhcp;
```